code/FsComposite/Composite.Simple/Data.fs

```diff
@@ -9,24 +11,75 @@ module Data =
         | A
         | B
         | C
+        | D
 
     //how to expand
 
-    let expand obj =
+    let expandSimple obj =
         match obj with
-        | A -> l B
-        | x -> l x
+        | A -> ll B
+        | x -> ll x
 
     //how to fold
 
-    let transformAB obj =
+    let find_and_transform_AB () =
+        
         let f1 = function | A -> true | _ -> false
         let f2 = function | B -> true | _ -> false
-        let f = function | (Some a, Some b) -> l C | (None, Some b) -> l C | _ -> LazyList.empty
-        find_2_and_transform f1 f2 f obj
+        
+        let try_find_A lst =
+            try
+                Some (LazyList.find f1 lst)
+            with
+            | :? System.Collections.Generic.KeyNotFoundException -> None
+            
+        let try_find_B lst =
+            try
+                Some (LazyList.find f2 lst)
+            with
+            | :? System.Collections.Generic.KeyNotFoundException -> None
 
-    let transformABStrict obj =
-        let f1 = function | A -> true | _ -> false
-        let f2 = function | B -> true | _ -> false
-        let f = fun (a, b) -> C
-        find_2_and_transform_strict f1 f2 f obj
+        let transform lst =
+            match try_find_A lst, try_find_B lst with
+            | Some a, Some b -> ll C
+            | None, Some b -> ll C
+            | _ -> LazyList.empty
+        
+        (([f1; f2] |> LazyList.ofList), LazyList.empty, transform)
+
+    let find_and_transform_BC () =
+        
+        let f3 = function | B -> true | _ -> false
+        let f4 = function | C -> true | _ -> false
+        
+        let try_find_B lst =
+            try
+                Some (LazyList.find f3 lst)
+            with
+            | :? System.Collections.Generic.KeyNotFoundException -> None
+            
+        let try_find_C lst =
+            try
+                Some (LazyList.find f4 lst)
+            with
+            | :? System.Collections.Generic.KeyNotFoundException -> None
+
+        let transform lst =
+            match try_find_B lst, try_find_C lst with
+            | Some b, Some c -> ll D
+            | None, Some C -> ll D
+            | _ -> LazyList.empty
+
+        (([f3; f4] |> LazyList.ofList), LazyList.empty, transform)

```

Comment:
precisely. the thing should work like this:
1) we have a step function that returns 
    - a list of test functions [is1; is2; is3]
    - a transform function that takes a list of arguments [a1; a2; a3] where a1 is an element satisfying is1, etc. and returns the list of results produced based on that list.

2) we can use the step algorithm specified this way like this:
    - when we analyze an element we are supplied with an accumulator [None; None; None]
    - we zip the given functions list with the accumulator list: fx, None -> if (fx x) then Some x else None
     - if the accumulator has no None elements in it then we can use the respective transform on it, add its product to the results and exclude the is-transform package from further results production process.